src/services/http.ts

```diff
@@ -0,0 +1,53 @@
+import log from 'electron-log/main'
+import { readFile } from 'node:fs/promises'
+import { getGlobalDispatcher, ProxyAgent, setGlobalDispatcher } from 'undici'
+
+import { UpstreamProxySettings } from '@/schemas/settings'
+
+const DEFAULT_DISPATCHER = getGlobalDispatcher()
+
+async function readCertificate(path: string | undefined) {
+  if (path === undefined) {
+    return undefined
+  }
+
+  try {
+    return await readFile(path, 'utf-8')
+  } catch (error) {
+    log.error('Failed to read proxy certificate.', error)
+
+    return undefined
+  }
+}
+
+function getAuthorizationHeader({ username, password }: UpstreamProxySettings) {
+  if (username === undefined || password === undefined) {
```

Comment:
The typing here is a bit unfortunate. The schema should have been:

```ts
const CredentialsSchema = z.object({
  username: z.string() // Not optional
  password: z.string() // Not optional
})

const UpstreamProxySettingsSchema = z.object({
  credentials: CredentialsSchema.optional() // Optional

  // no longer required.
  // requiresAuth: z.boolean()
})
```

But that would require a migration, so I'm biting the bullet.